packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -295,178 +157,234 @@ contract ControllerHelper is UniswapControllerHelper, AaveControllerHelper, IERC
+                    1e18
+                ),
+                uint8(ControllerHelperDataType.CALLBACK_SOURCE.SWAP_EXACTOUT_ETH_WPOWERPERP),
+                ""
+            );
+        } else if (_params.wPowerPerpAmountDesired < wPowerPerpAmountInLp) {
+            // if the new position target lower wPowerPerp amount, swap excess to WETH (position target higher WETH amount)
+            uint256 wPowerPerpExcess = wPowerPerpAmountInLp.sub(_params.wPowerPerpAmountDesired);
+
             _exactInFlashSwap(
                 wPowerPerp,
                 weth,
                 IUniswapV3Pool(wPowerPerpPool).fee(),
-                wPowerPerpAmount,
-                _params.limitPriceEthPerPowerPerp.mul(wPowerPerpAmount).div(1e18),
-                uint8(CALLBACK_SOURCE.SWAP_EXACTIN_WPOWERPERP_ETH),
+                wPowerPerpExcess,
+                _params.limitPriceEthPerPowerPerp.mul(wPowerPerpExcess).div(1e18),
+                uint8(ControllerHelperDataType.CALLBACK_SOURCE.SWAP_EXACTIN_WPOWERPERP_ETH),
                 ""
             );
         }
 
-        IWETH9(weth).withdraw(IWETH9(weth).balanceOf(address(this)));
-        payable(msg.sender).sendValue(address(this).balance);
+        // mint new position
+        ControllerHelperUtil.lpWPowerPerpPool(
+            controller,
+            nonfungiblePositionManager,
+            wPowerPerpPool,
+            wPowerPerp,
+            0,
+            ControllerHelperDataType.LpWPowerPerpPool({
+                recipient: msg.sender,
+                ethAmount: _params.ethAmountToLp,
+                amount0Desired: (isWethToken0) ? _params.wethAmountDesired : _params.wPowerPerpAmountDesired,
+                amount1Desired: (isWethToken0) ? _params.wPowerPerpAmountDesired : _params.wethAmountDesired,
+                amount0Min: _params.amount0DesiredMin,
+                amount1Min: _params.amount1DesiredMin,
+                lowerTick: _params.lowerTick,
+                upperTick: _params.upperTick
+            })
+        );
+
+        ControllerHelperUtil.sendBack(weth);
+    }
+
+    /**
+     * @notice Rebalance, increase and decrease LP liquidity through minting/burning wPowerPerp in vault
+     * @param _vaultId vault ID
+     * @param _collateralToFlashloan collateral amount to flashloan and deposit into vault to be able to withdraw Uni LP NFT
+     * @param _params array of ControllerHelperDataType.RebalanceVaultNftParams structs
+     */
+    function rebalanceVaultNft(
+        uint256 _vaultId,
+        uint256 _collateralToFlashloan,
+        ControllerHelperDataType.RebalanceVaultNftParams[] calldata _params
+    ) external payable {
+        _flashLoan(
+            weth,
+            _collateralToFlashloan,
+            uint8(ControllerHelperDataType.CALLBACK_SOURCE.FLASHLOAN_REBALANCE_VAULT_NFT),
+            abi.encode(_vaultId, _params)
+        );
     }
 
```

Comment:
If I have an LP with 100 eth and 0 oSQTH and I want to close it, sell 50 eth, and open a new LP with 50 eth and the oSQTH proceeds, do I need an extra action for rebalanceVaultNftType?

Reply:
Similarly, if I have LP with 100 eth and 30 oSQTH what would be the most natural way to:
- Sell 50% of the vault
- Swap 50 eth for oSQTH
- Burn 15 oSQTH and the proceeds of the swap against the vault

Reply:
I'm not sure I saw withdraw ETH => selling to oSQTH => mint in the functionalities doc ?

Reply:
Fair. Also on second thoughts it's just as easy to do any sale separately after the rebalance.